### PR TITLE
Reorganize registry deployment guide

### LINKED ISSUE #3433: Document pushing foreign layers
https://github.com/moby/moby/pull/33151

### PR DESCRIPTION
**Pull request by mstanleyjones**
### Proposed changes

Rewrites and reorganization
Also add information about pushing non-distributable layers
into a private registry

Fixes #3433 

### Unreleased project version (optional)

Docker 17.06

PTAL @stevvooe @dmcgowan @thaJeztah @cpuguy83

### CODE REVIEW

**Comment by mstanleyjones** [comment]
PTAL @friism

File: registry/deploying.md [hunk 1]
```diff
@@ -1,94 +1,240 @@
...
+[configuration reference](configuration.md).
 
-Start your registry:
+If you have an air-gapped datacenter, see
+[Considerations for air-gapped registries](#considerations-for-air-gapped-registries).
 
-    docker run -d -p 5000:5000 --restart=always --name registry registry:2
+## Run a local registry
+
+Use a command like the following to start the registry container:
```
Comment (thaJeztah) [inline registry/deploying.md]: We should point out that the example here is just for testing / trying out, because;

- it's insecure (there's no TLS, nor authentication in place, and the port is published, so anyone has access to your registry)
- there's no volume used, so images are stored on the container's filesystem; removing the container removes all your images

File: registry/deploying.md [hunk 2]
```diff
@@ -214,15 +394,68 @@ registry:
...
+
+- Certain images, such as the official Microsoft Windows base images, are not
+  distributable. This means that when you push an image based on one of these
+  images to your private registry, the non-distributable layers are **not**
+  pushed, but are always fetched from their authorized location. This is fine
+  for internet-connected hosts, but will not work in an air-gapped set-up.
+
+  In Docker 17.06 and higher, you can configure the Docker daemon to allow
+  pushing non-distributable layers to private registries, in this scenario.
+  **This is only useful in air-gapped set-ups in the presence of
```
Comment (friism) [inline registry/deploying.md]: _Strictly_ speaking, this is also useful in bandwidth-constrained setups, or setups where one wants to not be dependent on fetching the base layers from MS servers. I have no quarrel with the current formulation though.

File: registry/deploying.md [hunk 3]
```diff
@@ -214,15 +394,68 @@ registry:
...
+
+  3.  Restart the registry if it does not start automatically.
+
+  4.  When you push images to the registries in the list, their
+      non-distributable layers will be pushed to the registry.
+
+      >  **Warning**: Non-distributable artifacts typically have restrictions on
+      > how and where they can be distributed and shared. Only use this feature
+      > to push artifacts to private registries and ensure that you are in
+      > compliance with > any terms that cover redistributing non-distributable
```
Comment (friism) [inline registry/deploying.md]: bonus `>`

**Comment by mstanleyjones** [comment]
Addressed feedback.

File: docker-for-mac/networking.md [hunk 4]
```diff
@@ -78,11 +78,10 @@ There are two scenarios that the above limitations will affect:
 
 #### I want to connect from a container to a service on the host
 
-The Mac has a changing IP address (or none if you have no network access). Our
-current recommendation is to attach an unused IP to the `lo0` interface on the
-Mac; for example: `sudo ifconfig lo0 alias 10.200.10.1/24`, and make sure that
-your service is listening on this address or `0.0.0.0` (ie not `127.0.0.1`).
-Then containers can connect to this address.
+The Mac has a changing IP address (or none if you have no network access). From
```
Comment (stevvooe) [inline docker-for-mac/networking.md]: This should probably be clarified to OSX, so it is not confused with `MAC Address`.

File: engine/swarm/networking.md [hunk 5]
```diff
@@ -348,6 +348,41 @@ order to delete an existing bridge. The package name is `bridge-utils`.
 
 5.  Initialize or join the swarm.
 
+## Use a separate interface for control and data traffic
```
Comment (stevvooe) [inline engine/swarm/networking.md]: Is this related to registry deployment?

File: registry/deploying.md [hunk 6]
```diff
@@ -1,94 +1,246 @@
...
+  registry:2
+```
+
+## Storage customization
+
+### Customize the storage location
+
+By default, your registry data is persisted as a
+[docker volume](/engine/tutorials/dockervolumes.md) on the host filesystem.
+Just as any other container, you can use a bind mount instead. The following
```
Comment (stevvooe) [inline registry/deploying.md]: "a bind mount or volume instead"

File: registry/deploying.md [hunk 7]
```diff
@@ -1,94 +1,246 @@
...
+  --name registry \
+  -v `pwd`/data:/var/lib/registry \
+  registry:2
+```
 
-### Alternatives
+### Customize the storage back-end
 
-You should usually consider using [another storage backend](./storage-drivers/index.md) instead of the local filesystem. Use the [storage configuration options](./configuration.md#storage) to configure an alternate storage backend.
+Instead of the local filesystem, you can use a different storage back-end by
```
Comment (stevvooe) [inline registry/deploying.md]: You might want to compare this to usage of a volume.

File: registry/deploying.md [hunk 8]
```diff
@@ -1,94 +1,246 @@
...
+    ```
+
+3.  Restart the registry, directing it to use the TLS certificate. This command
+    bind-mounts the `certs/` directory into the container at `/certs/`, and sets
+    environment variables that tell the container where to find the `domain.crt`
+    and `domain.key` file.
+
+    ```bash
+    $ docker run -d \
+      -p 5000:5000 \
```
Comment (stevvooe) [inline registry/deploying.md]: Can we do this example without port 5000? The section above makes that promise then the example breaks that promise. ;)

File: registry/deploying.md [hunk 9]
```diff
@@ -1,94 +1,246 @@
...
+browser-trusted certificate. For more information on Let's Encrypt, see
+[https://letsencrypt.org/how-it-works/](https://letsencrypt.org/how-it-works/)
+and the relevant section of the
+[registry configuration](configuration.md#letsencrypt).
 
-### Alternatives
+### Use an insecure registry (testing only)
 
-While rarely advisable, you may want to use self-signed certificates instead, or use your registry in an insecure fashion. You will find instructions [here](insecure.md).
+It is possible to use a self-signed certificate, or to use our registry
```
Comment (stevvooe) [inline registry/deploying.md]: You can still setup verification with a self-signed certificate that isn't used for testing.

File: registry/deploying.md [hunk 10]
```diff
@@ -141,24 +294,45 @@ checks can be directed at "/", which will always return a `200 OK` response.
...
 
-The simplest way to achieve access restriction is through basic authentication (this is very similar to other web servers' basic authentication mechanism).
+The simplest way to achieve access restriction is through basic authentication
+(this is very similar to other web servers' basic authentication mechanism).
+This example uses native basic authentication using `htpasswd` to store the
+secrets.
 
 > **Warning**:
-> You **cannot** use authentication with an insecure registry. You have to [configure TLS first](deploying.md#running-a-domain-registry) for this to work.
+> You **cannot** use authentication with an insecure registry. You must
```
Comment (stevvooe) [inline registry/deploying.md]: ...with authentication schemes that send credentials as clear text.

**Comment by mstanleyjones** [comment]
@stevvooe the first two bits of feedback you had were for unrelated changes that just indicated this patch needed a rebase. Maybe @londoncalling can make note of them to be changed elsewhere.

File: registry/deploying.md (hunk 6 above)
Comment (mstanleyjones) [inline registry/deploying.md]: I don't think so. Line 134 already says it defaults to using a volume.

**Comment by mstanleyjones** [comment]
Addressed feedback and added a service example that uses secrets to store the TLS certificate material.

**Comment by mstanleyjones** [comment]
LGTM, merging into `vnext-engine`.
